window-copy.c

```diff
@@ -3097,27 +3134,94 @@ window_copy_search(struct window_mode_entry *wme, int direction, int regex,
 	wrapflag = options_get_number(wp->window->options, "wrap-search");
 	cis = window_copy_is_lowercase(str);
 
-	if (direction)
+	keys = options_get_number(wp->window->options, "mode-keys");
+
+	if (direction) {
+		/*
+		 * Behave according to mode-keys. If it is emacs, search forward
+		 * leaves the cursor after the match. If it is vi, the cursor
+		 * remains at the beginning of the match, regardless of
+		 * direction, which means that we need to start the next search
+		 * after the term the cursor is currently on when searching
+		 * forward.
+		 */
+		if (keys == MODEKEY_VI) {
+			if (data->searchmark != NULL)
+				window_copy_move_after_search_mark(data, &fx,
+				    &fy, wrapflag);
+			else {
+				/*
+				 * When there are no search marks, start the
+				 * search after the current cursor position.
+				 */
+				window_copy_move_right(s, &fx, &fy, wrapflag);
+			}
+		}
 		endline = gd->hsize + gd->sy - 1;
+	}
 	else {
-		if (again)
```

Comment:
It's unclear to me what this flag was for in this case. In what situations do we not want to backtrack before searching again?

Reply:
This made sense in the context of the old behaviour. In `master`, vi mode, if you have the pattern `xxx|b|bbxxxbbbbyyy` and the cursor is on the "piped" `|b|` and you search `?bbb` then the cursor won't move and you will get the `|b|` as the first match, since `again==0`. The `n`ext search will have `again==1` and will jump backward (and wrap). Since this is not what vim does (and now we do what vim does) we don't need this flag anymore.